src/observed/abstract.jl

```diff
@@ -8,3 +8,143 @@ Rows are samples, columns are observed variables.
+# reorder the columns of the data to match the speciation.
+# If no variable names are provided at all, generates the symbols in the form
+# Symbol(observed_var_prefix, i) for i=1:nobserved_vars.
+function prepare_data(
+    data::Union{AbstractDataFrame, AbstractMatrix, Nothing},
+    observed_vars::Union{AbstractVector, Nothing},
+    spec::Union{SemSpecification, Nothing},
+    nobserved_vars::Union{Integer, Nothing} = nothing;
+    observed_var_prefix::Union{Symbol, AbstractString},
+)
+    obs_vars = nothing
+    obs_vars_perm = nothing
+    if !isnothing(observed_vars)
+        obs_vars = Symbol.(observed_vars)
+        if !isnothing(spec)
+            obs_vars_spec = SEM.observed_vars(spec)
+            try
+                obs_vars_perm = source_to_dest_perm(
+                    obs_vars,
+                    obs_vars_spec,
+                    one_to_one = false,
+                    entities = "observed_vars",
+                )
+            catch err
+                if isa(err, KeyError)
+                    throw(
+                        ArgumentError(
+                            "observed_var \"$(err.key)\" from SEM specification is not listed in observed_vars argument",
+                        ),
+                    )
+                else
+                    rethrow(err)
+                end
+            end
+            # ignore trivial reorder
+            if obs_vars_perm == eachindex(obs_vars)
+                obs_vars_perm = nothing
+            end
+        end
+    elseif !isnothing(spec)
+        obs_vars = SEM.observed_vars(spec)
+    end
+    # observed vars in the order that matches the specification
+    obs_vars_reordered = isnothing(obs_vars_perm) ? obs_vars : obs_vars[obs_vars_perm]
+
+    # subset the data, check that obs_vars matches data or guess the obs_vars
+    if data isa AbstractDataFrame
+        if !isnothing(obs_vars_reordered) # subset/reorder columns
+            data = data[:, obs_vars_reordered]
+        else # default symbol names
+            obs_vars = obs_vars_reordered = Symbol.(names(data))
+        end
+        data_mtx = Matrix(data)
+    elseif data isa AbstractMatrix
+        if !isnothing(obs_vars)
+            size(data, 2) == length(obs_vars) || DimensionMismatch(
+                "The number of columns in the data matrix ($(size(data, 2))) does not match the length of observed_vars ($(length(obs_vars))).",
+            )
+            # reorder columns to match the spec
+            data_ordered = !isnothing(obs_vars_perm) ? data[:, obs_vars_perm] : data
+        else
+            obs_vars =
+                obs_vars_reordered =
+                    default_observed_vars(size(data, 2), observed_var_prefix)
+            data_ordered = data
+        end
+        # make sure data_mtx is a dense matrix (required for methods like mean_and_cov())
+        data_mtx =
+            data_ordered isa DenseMatrix ? data_ordered : convert(Matrix, data_ordered)

```

Comment:
It seems like the only subtypes of `DenseMatrix` are `Base.Experimental.Const{T, 2} where T` and `Matrix`, so maybe we can just do this:
```suggestion
        data_mtx = convert(Matrix, data_ordered)
```